File: app/javascript/controllers/booking_auth_controller.js
// app/javascript/controllers/booking_auth_controller.js
import { Controller } from "@hotwired/stimulus"
import { fetchWithAuth } from "../utils/fetch_helper"

/**
 * Booking Authentication Controller
 * Handles the dynamic booking authentication flow with minimal page reloads
 * @extends Controller
 */
export default class extends Controller {
    static targets = ["modal", "phoneSection", "emailSection", "otpSection", "errorMessage", "successMessage"]
    static values = {
        checkPhoneUrl: String,
        createUserUrl: String,
        verifyOtpUrl: String,
        resendOtpUrl: String,
        bookingParams: Object
    }

    /**
     * Initializes the controller
     * Sets up the modal and initial state
     */
    connect() {
        this.showModal()
        this.resetForm()
    }

    /**
     * Shows the authentication modal
     */
    showModal() {
        this.modalTarget.classList.remove('hidden')
    }

    /**
     * Hides the authentication modal
     */
    hideModal() {
        this.modalTarget.classList.add('hidden')
    }

    /**
     * Resets the form to its initial state
     */
    resetForm() {
        this.phoneSectionTarget.classList.remove('hidden')
        this.emailSectionTarget.classList.add('hidden')
        this.otpSectionTarget.classList.add('hidden')
        this.errorMessageTarget.classList.add('hidden')
        this.successMessageTarget.classList.add('hidden')
    }

    /**
     * Handles phone number submission
     * Checks if user exists with the provided phone number
     * @param {Event} event - The form submission event
     */
    async submitPhone(event) {
        event.preventDefault()

        const formData = new FormData(event.target)
        const phone = formData.get('phone')

        try {
            const response = await fetch(this.checkPhoneUrlValue, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRF-Token': document.querySelector('[name="csrf-token"]').content
                },
                body: JSON.stringify({ phone })
            })

            const data = await response.json()

            if (data.status === 'otp_sent') {
                // User exists, show OTP section
                this.userId = data.user_id
                this.phoneSectionTarget.classList.add('hidden')
                this.otpSectionTarget.classList.remove('hidden')
                this.showMessage('success', 'OTP sent to your phone and email.')
            } else if (data.status === 'email_required') {
                // New user, show email section
                this.phoneNumber = phone
                this.phoneSectionTarget.classList.add('hidden')
                this.emailSectionTarget.classList.remove('hidden')
            }
        } catch (error) {
            this.showMessage('error', 'An error occurred. Please try again.')
        }
    }

    /**
     * Handles email submission for new users
     * Creates a new user with phone and email
     * @param {Event} event - The form submission event
     */
    async submitEmail(event) {
        event.preventDefault()

        const formData = new FormData(event.target)
        const email = formData.get('email')

        try {
            const response = await fetch(this.createUserUrlValue, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRF-Token': document.querySelector('[name="csrf-token"]').content
                },
                body: JSON.stringify({
                    phone: this.phoneNumber,
                    email: email,
                    booking_params: this.bookingParamsValue
                })
            })

            const data = await response.json()

            if (data.status === 'otp_sent') {
                // User created, show OTP section
                this.userId = data.user_id
                this.emailSectionTarget.classList.add('hidden')
                this.otpSectionTarget.classList.remove('hidden')
                this.showMessage('success', 'OTP sent to your phone and email.')
            } else if (data.errors) {
                // Show validation errors
                this.showMessage('error', this.formatErrors(data.errors))
            }
        } catch (error) {
            this.showMessage('error', 'An error occurred. Please try again.')
        }
    }

    /**
     * Handles OTP verification
     * Verifies the OTP and completes the booking process
     * @param {Event} event - The form submission event
     */
    async verifyOtp(event) {
        event.preventDefault()

        const formData = new FormData(event.target)
        const otp = formData.get('otp')

        try {
            const response = await fetch(this.verifyOtpUrlValue, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRF-Token': document.querySelector('[name="csrf-token"]').content
                },
                body: JSON.stringify({
                    user_id: this.userId,
                    otp: otp
                })
            })

            const data = await response.json()

            if (data.status === 'success') {
                // OTP verified, complete booking
                this.completeBooking()
            } else {
                this.showMessage('error', 'Invalid OTP. Please try again.')
            }
        } catch (error) {
            this.showMessage('error', 'An error occurred. Please try again.')
        }
    }

    /**
     * Resends OTP to the user
     */
    async resendOtp() {
        try {
            const response = await fetch(this.resendOtpUrlValue, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRF-Token': document.querySelector('[name="csrf-token"]').content
                },
                body: JSON.stringify({ user_id: this.userId })
            })

            const data = await response.json()

            if (data.status === 'otp_resent') {
                this.showMessage('success', 'OTP has been resent.')
            }
        } catch (error) {
            this.showMessage('error', 'Failed to resend OTP. Please try again.')
        }
    }

    /**
     * Completes the booking process after successful authentication
     */
    async completeBooking() {
        try {
            // Create the booking with the stored parameters
            const response = await fetch('/bookings', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRF-Token': document.querySelector('[name="csrf-token"]').content
                },
                body: JSON.stringify({
                    booking: this.bookingParamsValue
                })
            })

            if (response.ok) {
                this.showMessage('success', 'Booking confirmed successfully!')
                // Redirect to booking details or show success message
                setTimeout(() => {
                    window.location.href = '/bookings'
                }, 2000)
            } else {
                this.showMessage('error', 'Failed to create booking. Please try again.')
            }
        } catch (error) {
            this.showMessage('error', 'An error occurred. Please try again.')
        }
    }

    /**
     * Displays a message to the user
     * @param {String} type - The type of message (success or error)
     * @param {String} message - The message to display
     */
    showMessage(type, message) {
        if (type === 'success') {
            this.successMessageTarget.textContent = message
            this.successMessageTarget.classList.remove('hidden')
            this.errorMessageTarget.classList.add('hidden')
        } else {
            this.errorMessageTarget.textContent = message
            this.errorMessageTarget.classList.remove('hidden')
            this.successMessageTarget.classList.add('hidden')
        }
    }

    /**
     * Formats error messages from the server
     * @param {Object} errors - The error object from the server
     * @returns {String} Formatted error message
     */
    formatErrors(errors) {
        return Object.entries(errors)
            .map(([field, messages]) => `${field} ${messages.join(', ')}`)
            .join('. ')
    }
}

File: app/controllers/booking_auth_controller.rb
# app/controllers/booking_auth_controller.rb
class BookingAuthController < ApplicationController
  skip_before_action :authenticate_user!

  # Displays the initial booking authentication form
  # @return [HTML] Renders the phone number entry form
  # @example Trigger: Visiting /booking/auth
  # @example Expected: Shows a form to enter phone number
  def new
    @booking = Booking.new
  end

  # Processes the phone number submission
  # Checks if a user exists with the provided phone number
  # @param :phone_number [String] The user's phone number
  # @return [Redirect] Redirects to OTP verification for existing users or email collection for new users
  # @example Trigger: Submitting the phone number form
  # @example Usage: POST /booking/auth with phone_number: "+1234567890"
  # @example Expected: Redirects based on user's existence
  def create
    phone_number = params[:phone_number]
    @user = User.find_by(phone_number: phone_number)

    if @user
      send_otp_to_phone(@user)
      redirect_to booking_otp_verify_path(phone_number: phone_number)
    else
      redirect_to booking_email_path(phone_number: phone_number)
    end
  end

  # Displays the email collection form for new users
  # @param :phone_number [String] The user's phone number from previous step
  # @return [HTML] Renders the email entry form
  # @example Trigger: Redirect from create action for new users
  # @example Expected: Shows a form to enter email address

  # todo: use stimilus for check whether the client exist and trigger otp send, let the create method ot be doing it
  def email
    @phone_number = params[:phone_number]
  end

  # Creates a new user account with phone number and email
  # @param phone_number [String] The user's phone number
  # @param email [String] The user's email address
  # @return [Redirect] Redirects to OTP verification on success, back to email form on failure
  # @example Trigger: Submitting the email form
  # @example Usage: POST /booking/email with phone_number: "+1234567890", email: "user@example.com"
  # @example Expected: Creates user account and redirects to OTP verification
  def create_user
    phone_number = params[:phone_number]
    email = params[:email]

    @user = User.new(
      phone_number: phone_number,
      email: email,
      password: Devise.friendly_token[0, 20]
    )

    if @user.save
      send_otp_to_email(@user)
      redirect_to booking_otp_verify_path(phone_number: phone_number)
    else
      flash[:error] = "Could not create account: #{@user.errors.full_messages.join(', ')}"
      redirect_to booking_email_path(phone_number: phone_number)
    end
  end

  # Displays the OTP verification form
  # @param phone_number [String] The user's phone number for OTP verification
  # @return [HTML] Renders the OTP entry form
  # @example Trigger: Redirect from create or create_user actions
  # @example Expected: Shows a form to enter OTP code
  def otp_verify
    @phone_number = params[:phone_number]
  end

  # Verifies the submitted OTP code
  # @param phone_number [String] The user's phone number
  # @param otp [String] The OTP code entered by the user
  # @return [Redirect] Signs in the user and redirects to booking on success, back to OTP form on failure
  # @example Trigger: Submitting the OTP form
  # @example Usage: POST /booking/otp_verify with phone_number: "+1234567890", otp: "123456"
  # @example Expected: Signs in user and redirects to booking or shows error
  def verify_otp
    phone_number = params[:phone_number]
    otp = params[:otp]

    @user = User.find_by(phone_number: phone_number)

    if @user && @user.validate_and_consume_otp!(otp)
      sign_in(@user, bypass: true)
      redirect_to new_booking_path, notice: "Successfully verified!"
    else
      flash[:error] = "Invalid OTP. Please try again."
      redirect_to booking_otp_verify_path(phone_number: phone_number)
    end
  end

  private

  # Sends OTP to user's phone number
  # @param user [User] The user to send OTP to
  # @return [Boolean] True if OTP was sent successfully
  # @example Trigger: Called from create action for existing users
  # @example Usage: send_otp_to_phone(user)
  # @example Expected: Sends SMS with OTP code
  def send_otp_to_phone(user)
    otp = user.generate_otp
    SmsService.send_otp(user.phone_number, otp)
  end

  # Sends OTP to user's email address
  # @param user [User] The user to send OTP to
  # @return [Boolean] True if email was sent successfully
  # @example Trigger: Called from create_user action for new users
  # @example Usage: send_otp_to_email(user)
  # @example Expected: Sends email with OTP code
  def send_otp_to_email(user)
    otp = user.generate_otp
    UserMailer.booking_otp(user, otp).deliver_later
  end
end